### PR TITLE
NodeMaterial: Basic BSDFs of MeshStandardMaterial and NodeBuilder simplification

### PR DESCRIPTION
**Pull request by sunag**
- [x] Basic BSDFs `MeshStandardMaterial` -> `roughnessNode`, `metalnessNode`,  `normalNode`
- [x] Simplification of `NodeBuilder` code generator and depedencies (biggest update of this PR)
- [x] `NormalMap` Support
- [x] 30+ math functions to `MathNode`
- [x] `ColorSpaceNode` all three.js encoding functions (automatic use in `TextureNode`)
- [x] Temporary variable system: `TempNode`
- [x] `StructNode` Support
- [x] New LightContext ( more generic )
- [x] `ExpressionNode` for inline code

*WebGPU - Selective Lights*
https://raw.githack.com/sunag/three.js/dev-nodematerial-igalia-1/examples/webgpu_lights_selective.html

image

```js

// PhysicalLightingModel example
// ReflectedLight use the same standard native of threejs

const glslCode = 
`void ( inout ReflectedLight reflectedLight, vec3 lightDirection, vec3 lightColor ) {

	RE_Direct_Physical( reflectedLight, lightDirection, lightColor );

}`

const physicalLightingModel = new FunctionNode(  glslCode   ).setIncludes( [ RE_Direct_Physical ] );
```

### CODE REVIEW

**Comment by sunag** [comment]
Sorry the big code!! The nexts PR will be smaller packages.

One question, I can remove `Phong Material` support to `WebGPU` in order to simplify?
This can be added by the user using `Custom lighting model`.

**Comment by Mugen87** [comment]
> One question, I can remove Phong Material support to WebGPU in order to simplify?

I would say yes.

**Comment by mrdoob** [comment]
Yay!

**Comment by mrdoob** [comment]
Hmm, I think there is something wrong with the normal map code though:

https://user-images.githubusercontent.com/97088/130590757-5df20bd1-35d8-4eda-acd7-eb24768592f4.mov

**Comment by sunag** [comment]
> Hmm, I think there is something wrong with the normal map code though:

Well noticied, fixed :) 

image

File: examples/jsm/renderers/nodes/core/ContextNode.js [hunk 1]
```diff
@@ -36,9 +36,11 @@ class ContextNode extends Node {
 
 	generate( builder, output ) {
 
+		const type = this.getType( builder );
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/ContextNode.js:39]: It seems `type` is not used in this method.

File: examples/jsm/renderers/nodes/core/StructNode.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		return new StructVarNode( this, inputs );
+
+	}
+
+	generate( builder, output ) {
+
+		const type = this.getType( builder );
+		const inputs = this.inputs;
+
+		const shaderStage = builder.getShaderStage();
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/StructNode.js:42]: Another unused variable.

File: examples/jsm/renderers/nodes/core/VaryNode.js [hunk 3]
```diff
@@ -23,14 +23,12 @@ class VaryNode extends Node {
...
-		// force nodeVary.snippet work in vertex stage
-		const snippet = this.value.buildStage( builder, NodeShaderStage.Vertex, type );
-
 		const nodeVary = builder.getVaryFromNode( this, type );
-		nodeVary.snippet = snippet;
-
 		const propertyName = builder.getPropertyName( nodeVary );
 
+		// force nodeVary.snippet work in vertex stage
+		const flowData = builder.flowNodeFromShaderStage( NodeShaderStage.Vertex, this.value, type, propertyName );
```
Comment (Mugen87) [inline examples/jsm/renderers/nodes/core/VaryNode.js:30]: And here, too.

**Comment by mrdoob** [comment]
Merging this. I think we can leave the unused vars clean up for another PR.

**Comment by mrdoob** [comment]
Thanks! Welcome back! 🤗

**Comment by sunag** [comment]
@mrdoob Thank you again!!! I am very happy to have returned. 🎉

**Pull request merged**